Mathlib/RingTheory/DedekindDomain/Different.lean

```diff
@@ -636,4 +646,109 @@ lemma pow_sub_one_dvd_differentIdeal [IsDedekindDomain A] [NoZeroSMulDivisors A
+    obtain ⟨y, hy, e⟩ := this
+    rw [← Algebra.algebraMap_intTrace (A := A), Algebra.linearMap_apply,
+      (IsLocalization.injective _ le_rfl).eq_iff] at e
+    exact e ▸ hy
+  refine FractionalIdeal.mul_induction_on (H ⟨_, hxQ, rfl⟩) ?_ ?_
+  · rintro x hx _ ⟨y, hy, rfl⟩
+    induction hy using Submodule.span_induction generalizing x with
+    | mem y h =>
+      obtain ⟨y, hy, rfl⟩ := h
+      obtain ⟨z, hz⟩ :=
+        (FractionalIdeal.mem_dual (by simp)).mp hx 1 ⟨1, trivial, (algebraMap B L).map_one⟩
+      simp only [Algebra.traceForm_apply, mul_one] at hz
+      refine ⟨z * y, Ideal.mul_mem_left _ _ hy, ?_⟩
+      rw [Algebra.linearMap_apply, Algebra.linearMap_apply, mul_comm x,
+        ← IsScalarTower.algebraMap_apply,
+        ← Algebra.smul_def, LinearMap.map_smul_of_tower, ← hz,
+        Algebra.smul_def, map_mul, mul_comm]
+    | zero => simp
+    | add y z _ _ hy hz =>
+      simp only [map_add, mul_add]
+      exact Submodule.add_mem _ (hy x hx) (hz x hx)
+    | smul y z hz IH =>
+      simpa [Algebra.smul_def, mul_assoc, -FractionalIdeal.mem_coeIdeal, mul_left_comm x] using
+        IH _ (Submodule.smul_mem _ y hx)
+  · simp only [map_add, forall_exists_index, and_imp]
+    exact fun _ _ h₁ h₂ ↦ Submodule.add_mem _ h₁ h₂
+
+open nonZeroDivisors
+
+omit [IsIntegrallyClosed A] in
+theorem not_dvd_differentIdeal_of_isCoprime_of_isSeparable
+    [IsDedekindDomain A] [NoZeroSMulDivisors A B] [Module.Finite A B]
+    [Algebra.IsSeparable (FractionRing A) (FractionRing B)]
+    {p : Ideal A} [p.IsMaximal] (P Q : Ideal B) [P.IsMaximal] [P.LiesOver p]
+    (hPQ : IsCoprime P Q) (hP : P * Q = Ideal.map (algebraMap A B) p)
+    [Algebra.IsSeparable (A ⧸ p) (B ⧸ P)] :
+    ¬ P ∣ differentIdeal A B := by
+  letI : Algebra (A ⧸ p) (B ⧸ Q) := Ideal.Quotient.algebraQuotientOfLEComap (by
+      rw [← Ideal.map_le_iff_le_comap, ← hP]
+      exact Ideal.mul_le_left)
+  have : IsScalarTower A (A ⧸ p) (B ⧸ Q) := .of_algebraMap_eq' rfl
+  have : Module.Finite (A ⧸ p) (B ⧸ Q) :=
+    Module.Finite.of_restrictScalars_finite A (A ⧸ p) (B ⧸ Q)
+  letI e : (B ⧸ p.map (algebraMap A B)) ≃ₐ[A ⧸ p] ((B ⧸ P) × B ⧸ Q) :=
+    { __ := (Ideal.quotEquivOfEq hP.symm).trans (Ideal.quotientMulEquivQuotientProd P Q hPQ),
+      commutes' := Quotient.ind fun _ ↦ rfl }
+  obtain ⟨x, hx⟩ : ∃ x, Algebra.trace (A ⧸ p) (B ⧸ P) x ≠ 0 := by
+    simpa [LinearMap.ext_iff] using Algebra.trace_ne_zero (A ⧸ p) (B ⧸ P)
+  obtain ⟨y, hy⟩ := Ideal.Quotient.mk_surjective (e.symm (x, 0))
+  refine not_dvd_differentIdeal_of_intTrace_not_mem A P Q hP y ?_ ?_
+  · simpa [e, Ideal.Quotient.eq_zero_iff_mem] using congr((e $hy).2)
+  · rw [← Ideal.Quotient.eq_zero_iff_mem, ← Algebra.trace_quotient_eq_of_isDedekindDomain,
+      hy, Algebra.trace_eq_of_algEquiv, Algebra.trace_prod_apply]
+    simpa
+
+omit [IsIntegrallyClosed A] in
+theorem not_dvd_differentIdeal_of_isCoprime
+    [IsDedekindDomain A] [NoZeroSMulDivisors A B] [Module.Finite A B]
```

Comment:
The last three theorem have a lot of common variables. Can you use a `variable` line?